imath removal, removal of non-combined modules (VMR, TE, TC, PR, ME, MP), and new combined TB module

aryd opened the pull request:
#### PR description:

This PR has code developed over a long time that includes the removal of the iMath code, removal of the non-combined modules and new TB code that combines TB for multiple seeding combinations. Plus many other cleanups.

#### PR validation:

This code is running and producing tracks. Need to check that test vectors are OK with HLS.

number of events = 100

efficiency for |eta| < 1.0 = 92.33 +- 0.76
efficiency for 1.0 < |eta| < 1.75 = 93.96 +- 0.93
efficiency for 1.75 < |eta| < 2.40 = 94.51 +- 1.26
combined efficiency for |eta| < 2.40 = 93.14 +- 0.54 = 2076.00/2229.00

efficiency for pt > 2.00 = 93.14 +- 0.54
efficiency for 2.00 < pt < 8.0 = 92.79 +- 0.63
efficiency for pt > 8.0 = 94.18 +- 1.00
efficiency for pt > 40.0 = 85.71 +- 5.91

 TP/event (pt > 2.00) = 156.71
 TP/event (pt > 3.0) = 52.16
 TP/event (pt > 10.0) = 4.63
 tracks/event (no pt cut)= 173.70
 tracks/event (pt > 2.00) = 162.58
 tracks/event (pt > 3.0) = 57.50
 tracks/event (pt > 10.0) = 5.93

Percentage fake tracks (pt > 2.00) = 2.20%
Percentage duplicate tracks (pt > 2.00)= 1.63%

z0 resolution = 0.10cm at |eta| = 0.05
z0 resolution = 0.41cm at |eta| = 1.95


N.B. Both this PR and https://github.com/cms-L1TK/cmssw/pull/312 modify the TrackletConfigBuilder, so there may be some conflict between them.

aryd committed:
squashed my commits

aryd committed:
Some fixes after rebase to get code to run

aryd committed:
Remove now usused combined flag in Settings.h

aryd committed:
code formatting

aryd committed:
Some more cleanup

aryd committed:
Fix inefficiency for hybrid tracks

aryd committed:
Cleanup/simplify code in FitTrack

aryd committed:
code-format

aryd committed:
Rewrite the filling of the Track and Stub Streams objects to work with two track builders

aryd committed:
run code-format

aryd committed:
First stab at a new configuration of the displaced tracking that uses the PC and VMSMER modules

aryd committed:
Further updates to the displaced tracking configuration. Now finds tracks

aehart committed:
Fix for displaced tracking in #304 (#306)

* Removed superfluous continue.

* Switch to Stub::rvalue().

* Switch to Stub::rvalue().

tschuh commented:
I had a look at the code generating the track and stub streams and it looks like you are messing up the stub channel assignment per seed type. That may lead to stub losses (seed type 3 projects only in 4 layers, stubs send on channel >6 won't be used) and wrong interpretation of stub data (e.g. ps barrel stub data interpreted as ps endcap stub). I would recommend to make the ChannelAssigment class available in TrackFit and to use it to make the correct assignments.

aehart committed:
More fixes for displaced tracking in #304 (#309)

* Added back triplet seeds to the seed map for now.

* Added separate FM memories for triplet seeds for now.

* Consider TPROJ memories unpaged.

* Added warnings when no projection memory exists.

* Replaced magic numbers with seed enum.

* Added an N_SEED_DISPLACED constant.

* Added wiring based on latest baseline wiring.

* Removed obsolete settings and methods.

aryd committed:
Fix radial position

aryd committed:
run code formatting

aryd committed:
Add comment

aryd committed:
Run code format

aryd committed:
Restore code

aryd committed:
Code format

aryd committed:
Fix radial position

aryd committed:
Remove printout

aryd committed:
Fix to handle that output tracks are not in the seed order

aryd commented:
After spending considerable time trying to understand the data organization in the ChannelAssignment, TrackStream, and StubStream classes I convinced myself that there was no problem with how was filling this in FitTrack.cc and Sector.cc. There were two issues; the first was easy, we had change the radial position to be stored in 11 bits using an offset. However, the KF code expects the corrected radial position that uses 12 bits. The issue that I struggled a lot with resolving was that in the L1FPGATrackProducer where the StreamsTrack object is filled it assumed that the tracks were stored in the track collection in order of the seed number. But as we now find tracks using two track builders that finds tracks from different seeds this is no longer the case. I wrote a patch that fixes this. But ultimately I think this should be fixed in some cleaner way.

The CI integration test is successful now - I hope we can merge this soon.

aehart committed:
Fixed style issues.

aehart commented:
Just FYI, I fixed some minor style issues, but because of issues with the Docker container, it took several retries to get the CI tests to run. This is why GitHub shows the tests as failing, although they did all pass once they were able to run:
https://gitlab.cern.ch/cms-l1tk/cmssw_CI/-/pipelines/10939129

aryd committed:
Updates to support reduced, L5L6, configuration

tomalin commented:
I've now fixed the "docker" problem that was causing git CI to crash. Please make a trivial one line change to this PR, to force a rerun of the CI.

tomalin commented:
Is this PR readding the extended wiring map files to TrackFindingTracklet/data/ after Claire Savard removed them all (by calculating the map inside TrackletConfigBuilder.cc)? Why? Does this PR need rebasing to include her changes?

tomalin commented:
I've checked the code style. Looks good.

aryd committed:
Update L1TrackNtupleMaker_cfg.py

Minimal change to force CI to rerun.

aryd commented:
Done.

 -Anders

Anders Ryd
***@***.******@***.***>



On Mar 21, 2025, at 12:37 PM, Ian Tomalin ***@***.******@***.***>> wrote:


I've now fixed the "docker" problem that was causing git CI to crash. Please make a trivial one line change to this PR, to force a rerun of the CI.

—
Reply to this email directly, view it on GitHub<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2743896131>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABI4LTNA57V6RPMPSHCQ3AD2VQ53PAVCNFSM6AAAAABURPQHKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBTHA4TMMJTGE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

[tomalin]tomalin left a comment (cms-L1TK/cmssw#304)<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2743896131>

I've now fixed the "docker" problem that was causing git CI to crash. Please make a trivial one line change to this PR, to force a rerun of the CI.

—
Reply to this email directly, view it on GitHub<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2743896131>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABI4LTNA57V6RPMPSHCQ3AD2VQ53PAVCNFSM6AAAAABURPQHKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBTHA4TMMJTGE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

aryd commented:
I’m not sure if this is a question for me. I did add files to TrackFindingTracklet/data for the reduced (L5L6) configuration. But I did not change anything related to the extended wiring map.

 -Anders


Anders Ryd
***@***.******@***.***>



On Mar 21, 2025, at 12:50 PM, Ian Tomalin ***@***.******@***.***>> wrote:


Is this PR readding the extended wiring map files to TrackFindingTracklet/data/ after Claire Savard removed them all (by calculating the map inside TrackletConfigBuilder.cc<http://trackletconfigbuilder.cc/>)? Why? Does this PR need rebasing to include her changes?

—
Reply to this email directly, view it on GitHub<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2743925184>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABI4LTNLYS7GJL6ZKIZIVRD2VQ7NHAVCNFSM6AAAAABURPQHKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBTHEZDKMJYGQ>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

[tomalin]tomalin left a comment (cms-L1TK/cmssw#304)<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2743925184>

Is this PR readding the extended wiring map files to TrackFindingTracklet/data/ after Claire Savard removed them all (by calculating the map inside TrackletConfigBuilder.cc<http://trackletconfigbuilder.cc/>)? Why? Does this PR need rebasing to include her changes?

—
Reply to this email directly, view it on GitHub<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2743925184>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABI4LTNLYS7GJL6ZKIZIVRD2VQ7NHAVCNFSM6AAAAABURPQHKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBTHEZDKMJYGQ>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

tomalin commented:
In the default branch, the data directory is empty, aside from the seedWiring.json file https://github.com/cms-L1TK/cmssw/tree/L1TK-dev-14_2_0_pre4/L1Trigger/TrackFindingTracklet/data .  But in your PR branch, this directory contains 6 additional files https://github.com/cms-L1TK/cmssw/tree/MultipleMPs_Rebase2_squashed/L1Trigger/TrackFindingTracklet/data . For example, it contains:
1) wires_hourglassExtendedAllCombined.dat
2) wires_reduced.dat
Why have these been added back? You can see that Claire Savard's PR https://github.com/cms-L1TK/cmssw/pull/312/files deleted wires_hourglassExtendedAllCombined.dat only a few weeks ago. If you want to modify her wiring map, you presumably need to edit TrackletConfigBuilder.cc .

aryd commented:
Ian,

 I see the following files:

[cid:DE308908-BFAA-447D-ABA6-0158D0D8DB38]

The *_reduced.dat were added by me to support the reduced L5L6 only configuration and I expect these to be there. However, I don’t know about the *hourglassExtended*.dat files… Maybe I misunderstand git, but it seems these files were added by Andrew ~2 weeks ago?

 -Anders


Anders Ryd
***@***.******@***.***>



On Mar 21, 2025, at 6:46 PM, Ian Tomalin ***@***.******@***.***>> wrote:


In the default branch, the data directory is empty, aside from the seedWiring.json file https://github.com/cms-L1TK/cmssw/tree/L1TK-dev-14_2_0_pre4/L1Trigger/TrackFindingTracklet/data . But in your PR branch, this directory contains 6 additional files https://github.com/cms-L1TK/cmssw/tree/MultipleMPs_Rebase2_squashed/L1Trigger/TrackFindingTracklet/data . For example, it contains:

  1.  wires_hourglassExtendedAllCombined.dat
  2.  wires_reduced.dat
Why have these been added back? You can see that Claire Savard's PR https://github.com/cms-L1TK/cmssw/pull/312/files deleted wires_hourglassExtendedAllCombined.dat only a few weeks ago. If you want to modify her wiring map, you presumably need to edit TrackletConfigBuilder.cc<http://trackletconfigbuilder.cc/> .

—
Reply to this email directly, view it on GitHub<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2744616163>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABI4LTKDL3QIQJASU52AHM32VSJDZAVCNFSM6AAAAABURPQHKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBUGYYTMMJWGM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

[tomalin]tomalin left a comment (cms-L1TK/cmssw#304)<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2744616163>

In the default branch, the data directory is empty, aside from the seedWiring.json file https://github.com/cms-L1TK/cmssw/tree/L1TK-dev-14_2_0_pre4/L1Trigger/TrackFindingTracklet/data . But in your PR branch, this directory contains 6 additional files https://github.com/cms-L1TK/cmssw/tree/MultipleMPs_Rebase2_squashed/L1Trigger/TrackFindingTracklet/data . For example, it contains:

  1.  wires_hourglassExtendedAllCombined.dat
  2.  wires_reduced.dat
Why have these been added back? You can see that Claire Savard's PR https://github.com/cms-L1TK/cmssw/pull/312/files deleted wires_hourglassExtendedAllCombined.dat only a few weeks ago. If you want to modify her wiring map, you presumably need to edit TrackletConfigBuilder.cc<http://trackletconfigbuilder.cc/> .

—
Reply to this email directly, view it on GitHub<https://github.com/cms-L1TK/cmssw/pull/304#issuecomment-2744616163>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABI4LTKDL3QIQJASU52AHM32VSJDZAVCNFSM6AAAAABURPQHKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBUGYYTMMJWGM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

tomalin commented:
@aehart can you clarify this please. As far as I can tell, the PR https://github.com/cms-L1TK/cmssw/pull/309 massively modified the file wires_hourglassExtendedAllCombined.dat . This PR was merged about two weeks ago. And then Claire's PR https://github.com/cms-L1TK/cmssw/pull/312/files deleted this file, because she moved the calculation of the extended wiring map into TrackletConfigBuilder.cc . I believe her PR was merged slightly after the first one, which is why the development branch https://github.com/cms-L1TK/cmssw/tree/L1TK-dev-14_2_0_pre4/L1Trigger/TrackFindingTracklet/data no longer contains wires_hourglassExtendedAllCombined.dat . But now this PR of Anders's is adding wires_hourglassExtendedAllCombined.dat back again, as one can see from https://github.com/cms-L1TK/cmssw/pull/304/files#diff-c2bdb1020463f45ab523cc85769d38f07210a3bbbe843ffcb943367cf49e8d68 . I believe this is a mistake. But as you were more closely involved in the previous PRs than me, you'll know better?

aehart committed:
Removed extended wiring made obsolete by #297.

aehart commented:
> @aehart can you clarify this please. As far as I can tell, the PR #309 massively modified the file wires_hourglassExtendedAllCombined.dat . This PR was merged about two weeks ago. And then Claire's PR https://github.com/cms-L1TK/cmssw/pull/312/files deleted this file, because she moved the calculation of the extended wiring map into TrackletConfigBuilder.cc . I believe her PR was merged slightly after the first one, which is why the development branch https://github.com/cms-L1TK/cmssw/tree/L1TK-dev-14_2_0_pre4/L1Trigger/TrackFindingTracklet/data no longer contains wires_hourglassExtendedAllCombined.dat . But now this PR of Anders's is adding wires_hourglassExtendedAllCombined.dat back again, as one can see from https://github.com/cms-L1TK/cmssw/pull/304/files#diff-c2bdb1020463f45ab523cc85769d38f07210a3bbbe843ffcb943367cf49e8d68 . I believe this is a mistake. But as you were more closely involved in the previous PRs than me, you'll know better?

The extended wires files should be completely superseded by the changes in #312, so I have removed them in the latest commit. They were being ignored already, so this should have no effect on the results.

tomalin reviewed (approved):
Approved
